File: Medium/98.Valid BST/solution.cpp
class Solution {
public:
    void check(bool &firsttime, TreeNode* root, int &prev, bool &ans) {
        if (!ans) return;
        if (!root) return;
        check(firsttime, root->left, prev, ans);
        if (root->val <= prev && !firsttime) {ans = false; return;}
        else {prev = root->val; firsttime = false;}
        check(firsttime, root->right, prev, ans);
    }
    bool isValidBST(TreeNode* root) {
        int prev = INT_MIN;
        bool firsttime = true;
        bool ans = true;
        check(firsttime, root, prev, ans);
        return ans;
    }
};

File: Medium/98.Valid BST/solution.java
import java.util.*;
public class isValidBST {

  static class Node{
    int data;
    Node left;
    Node right;
  

  Node(int data){
    this.data = data;
  }
  }



 public static boolean isValid(Node root , Node min , Node max){

      if(root == null){
        return true;
      }

      if(min!=null &&root.data <=min.data ){
        return false;
      }

      if(max !=null && root.data >=max.data ){
        return false;
      }
      return isValid(root.left,min,max) && isValid(root.right,root,max);
    }




public static void main(String args[]) {
 if(isValid(null ,null ,null)){
    System.out.println("Valid");
  }else{
    System.out.println("Not Valid");

  }
}
}